components/formats-api/src/loci/formats/WrappedReader.java

```diff
@@ -0,0 +1,473 @@
+/*
+ * #%L
+ * BSD implementations of Bio-Formats readers and writers
+ * %%
+ * Copyright (C) 2005 - 2017 Open Microscopy Environment:
+ *   - Board of Regents of the University of Wisconsin-Madison
+ *   - Glencoe Software, Inc.
+ *   - University of Dundee
+ * %%
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ * 
+ * 1. Redistributions of source code must retain the above copyright notice,
+ *    this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ * 
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ * #L%
+ */
+
+package loci.formats;
+
+import java.io.IOException;
+import java.util.Hashtable;
+import java.util.List;
+
+/**
+ * Helper class for readers which wrap other readers.
+ *
+ * Some methods can only be called before initFile/setId, others can only be called after.
+ * This class assumes the helper object is only available after initFile() has been called.
+ *
+ * Instead the inherited FormatReader and IMetadataConfigurable methods are kept since the properties
+ * they set are inherited. These properties can be set on the helper class in initFile() by calling
+ * callDeferredSetters().
+ *
+ * The following setters and corresponding getters are not passed to the helper.
+ * If you initialise the helper in the constructor instead of at runtime during initFile()
+ * you may wish to override them.
```

Comment:
You would override them to do the forwarding? Does callDeferredSetters help? Would a flag to detect where the helper was set help?

Reply:
It depends what the helper does. If the helper doesn't change its state anything between the construction of `WrappedReader` and `WrappedReader.initFile()` then either would work, in which case you might as well create the helper in `initFile()` (see second suggested improvement in the PR description). If it does change it's state then there's nothing we can do, it's down to the implementer to handle everything.